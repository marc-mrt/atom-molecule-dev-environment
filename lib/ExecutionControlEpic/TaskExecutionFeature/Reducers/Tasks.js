"use babel";
// @flow

import type { Task } from "../Types/types.js.flow";
import { Record, Map } from "immutable";

export default function(
  state: TasksReducer = new TasksReducerRecord(),
  action: any,
): TasksReducer {
  switch (action.type) {
    case "ADD_TASK":
      return state
        .update("byTaskID", byTaskID =>
          byTaskID.set(action.payload.task.id, action.payload.task),
        )
        .update("byPlanID", byPlanID =>
          byPlanID.set(action.payload.task.id, action.payload.task.plan.id),
        )
        .update("byToolID", byToolID =>
          byToolID.set(
            action.payload.task.id,
            action.payload.task.plan.tool.id,
          ),
        );
    case "START_TASK":
      return state.update(
        "byTaskID",
        byTaskID =>
          byTaskID.has(action.payload.id)
            ? state.byTaskID.update(
                action.payload.id,
                task =>
                  task.state == "created"
                    ? { ...task, state: "running", end: action.payload.date }
                    : task,
              )
            : byTaskID,
      );
    case "STOP_TASK":
      return state.update(
        "byTaskID",
        byTaskID =>
          byTaskID.has(action.payload.id)
            ? state.byTaskID.update(
                action.payload.id,
                task =>
                  task.state == "running"
                    ? { ...task, state: "stopped", end: action.payload.date }
                    : task,
              )
            : byTaskID,
      );
    case "KILL_TASK":
    case "STOP_TASK_OF_PLAN":
      return state.update("byTaskID", byTaskID =>
        byTaskID.map(
          task =>
            task.state == "running" &&
            state.byPlanID.get(task.id) == action.payload.plan.id
              ? { ...task, state: "stopped", end: action.payload.date }
              : task,
        ),
      );
    case "CRASH_TASK":
      return state.update(
        "byTaskID",
        byTaskID =>
          byTaskID.has(action.payload.id)
            ? state.byTaskID.update(
                action.payload.id,
                task =>
                  task.state == "running"
                    ? { ...task, state: "crashed", end: action.payload.date }
                    : task,
              )
            : byTaskID,
      );
    case "FAIL_TASK":
      return state.update(
        "byTaskID",
        byTaskID =>
          byTaskID.has(action.payload.id)
            ? state.byTaskID.update(
                action.payload.id,
                task =>
                  task.state == "running"
                    ? { ...task, state: "failed", end: action.payload.date }
                    : task,
              )
            : byTaskID,
      );
    case "SUCCEED_TASK":
      return state.update(
        "byTaskID",
        byTaskID =>
          byTaskID.has(action.payload.id)
            ? state.byTaskID.update(
                action.payload.id,
                task =>
                  task.state == "running"
                    ? { ...task, state: "succeed", end: action.payload.date }
                    : task,
              )
            : byTaskID,
      );
    case "NEXT_STEP":
      return state.update(
        "byTaskID",
        byTaskID =>
          byTaskID.has(action.payload.id)
            ? state.byTaskID.update(action.payload.id, task => ({
                ...task,
                step: task.step + 1,
              }))
            : byTaskID,
      );
    case "BUSY_TASK":
      return state.update(
        "byTaskID",
        byTaskID =>
          byTaskID.has(action.payload.id)
            ? state.byTaskID.update(action.payload.id, task => ({
                ...task,
                busy: true,
              }))
            : byTaskID,
      );
    case "WAITING_TASK":
      return state.update(
        "byTaskID",
        byTaskID =>
          byTaskID.has(action.payload.id)
            ? state.byTaskID.update(action.payload.id, task => ({
                ...task,
                busy: false,
              }))
            : byTaskID,
      );
    case "TERMINAL_INITIALIZED":
      return state.update(
        "byTaskID",
        byTaskID =>
          byTaskID.has(action.payload.id)
            ? state.byTaskID.update(action.payload.id, task => ({
                ...task,
                terminal: true,
              }))
            : byTaskID,
      );
    default:
      return state;
  }
}

export const TasksReducerRecord = Record({
  byTaskID: Map(),
  byPlanID: Map(),
  byToolID: Map(),
});

export type TasksReducer = {
  byTaskID: Map<string, Task>,
  byPlanID: Map<string, string>,
  byToolID: Map<string, string>,
};
