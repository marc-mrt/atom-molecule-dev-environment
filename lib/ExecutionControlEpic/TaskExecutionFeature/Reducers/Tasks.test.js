"use babel";
// @flow

import type { TasksReducer } from "./Tasks";
import Tasks from "./Tasks";
import type { AddTaskAction } from "../Actions/AddTask";
import { addTask } from "../Actions/AddTask";
import type { CrashTaskAction } from "../Actions/CrashTask";
import { crashTask } from "../Actions/CrashTask";
import type { StopTaskAction } from "../Actions/StopTask";
import { stopTask } from "../Actions/StopTask";
import type { StartTaskAction } from "../Actions/StartTask";
import { startTask } from "../Actions/StartTask";
import type { SucceedTaskAction } from "../Actions/SucceedTask";
import { succeedTask } from "../Actions/SucceedTask";
import type { FailTaskAction } from "../Actions/FailTask";
import { failTask } from "../Actions/FailTask";
import type { StopTaskOfPlanAction } from "../Actions/StopTaskOfPlan";
import { stopTaskOfPlan } from "../Actions/StopTaskOfPlan";
import type { BusyTaskAction } from "../Actions/BusyTask";
import { busyTask } from "../Actions/BusyTask";
import type { WaitingTaskAction } from "../Actions/WaitingTask";
import { waitingTask } from "../Actions/WaitingTask";

describe("Tasks Reducer", () => {
  describe("ADD_TASK", () => {
    it("should add a task", () => {
      let action: AddTaskAction = addTask(
        "1",
        {
          config: 2,
          packageInfo: {
            name: "root",
            type: "file",
            path: "/",
          },
          stager: { type: "integrated", host: null },
          tool: { id: "3", name: "tool", iconUri: "" },
        },
        { type: "shell", command: "ls", cwd: "/etc" },
        1475954335,
      );
      let state: TasksReducer = [];
      let subject = Tasks(state, action);

      expect(subject).toEqual([action.payload.task]);
    });
  });

  describe("START_TASK", () => {
    it("should mark the task as started", () => {
      let action: StartTaskAction = startTask("2");
      let state: TasksReducer = [
        {
          id: "1",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "2",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "3",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
      ];
      let subject = Tasks(state, action);

      expect(subject).toMatchSnapshot();
    });
  });

  describe("STOP_TASK", () => {
    it("should mark the task as stopped", () => {
      let action: StopTaskAction = stopTask("2", 1475954336);
      let state: TasksReducer = [
        {
          id: "1",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "/ok" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "2",
          state: "running",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "/ok" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "3",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "/ok" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
      ];
      let subject = Tasks(state, action);

      expect(subject).toMatchSnapshot();
    });
  });

  describe("STOP_TASK_OF_PLAN", () => {
    it("should mark the task as stopped", () => {
      let action: StopTaskOfPlanAction = stopTaskOfPlan(
        {
          name: "my Plan",
          config: 2,
          packageInfo: {
            name: "root",
            type: "file",
            path: "/",
          },
          stager: { type: "integrated", host: null },
          tool: { id: "2", name: "tool", iconUri: "" },
          state: "running",
        },
        1475954336,
      );
      let state: TasksReducer = [
        {
          id: "1",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "2",
          state: "running",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "2",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "3",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
      ];
      let subject = Tasks(state, action);

      expect(subject).toMatchSnapshot();
    });
  });

  describe("CRASH_TASK", () => {
    it("should mark the task as crashed", () => {
      let action: CrashTaskAction = crashTask("2", 1475954336);
      let state: TasksReducer = [
        {
          id: "1",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "2",
          state: "running",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "3",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
      ];
      let subject = Tasks(state, action);

      expect(subject).toMatchSnapshot();
    });
  });

  describe("SUCCEED_TASK", () => {
    it("should mark the task as succeed", () => {
      let action: SucceedTaskAction = succeedTask("2", 1475954336);
      let state: TasksReducer = [
        {
          id: "1",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "2",
          state: "running",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "3",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
      ];
      let subject = Tasks(state, action);

      expect(subject).toMatchSnapshot();
    });
  });

  describe("FAIL_TASK", () => {
    it("should mark the task as failed", () => {
      let action: FailTaskAction = failTask("2", 1475954336);
      let state: TasksReducer = [
        {
          id: "1",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "2",
          state: "running",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "3",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
      ];
      let subject = Tasks(state, action);

      expect(subject).toMatchSnapshot();
    });
  });

  describe("BUSY_TASK", () => {
    it("should mark the task as busy", () => {
      let action: BusyTaskAction = busyTask("2");
      let state: TasksReducer = [
        {
          id: "1",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "2",
          state: "running",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "3",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
      ];
      let subject = Tasks(state, action);

      expect(subject).toMatchSnapshot();
    });
  });

  describe("WAITING_TASK", () => {
    it("should mark the task as waiting", () => {
      let action: WaitingTaskAction = waitingTask("2");
      let state: TasksReducer = [
        {
          id: "1",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
        {
          id: "2",
          state: "running",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
          busy: false,
        },
        {
          id: "3",
          state: "created",
          debut: 1475954335,
          strategy: { type: "shell", command: "", cwd: "" },
          plan: {
            config: 2,
            packageInfo: {
              name: "root",
              type: "file",
              path: "/",
            },
            stager: { type: "integrated", host: null },
            tool: {
              id: "1",
              name: "tool",
              iconUri: "",
            },
          },
          step: 1,
        },
      ];
      let subject = Tasks(state, action);

      expect(subject).toMatchSnapshot();
    });
  });
});
