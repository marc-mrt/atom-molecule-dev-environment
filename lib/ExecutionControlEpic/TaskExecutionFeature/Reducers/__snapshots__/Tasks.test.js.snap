// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks Reducer BUSY_TASK should mark the task as busy 1`] = `
Immutable.Record {
  "byTaskID": Immutable.Map {
    "1": Object {
      "debut": 1475954335,
      "id": "1",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "2": Object {
      "busy": true,
      "debut": 1475954335,
      "id": "2",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "running",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "3": Object {
      "debut": 1475954335,
      "id": "3",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
  },
  "byPlanID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
  "byToolID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
}
`;

exports[`Tasks Reducer CRASH_TASK should mark the task as crashed 1`] = `
Immutable.Record {
  "byTaskID": Immutable.Map {
    "1": Object {
      "debut": 1475954335,
      "id": "1",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "2": Object {
      "busy": false,
      "debut": 1475954335,
      "end": 1475954336,
      "id": "2",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "crashed",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "3": Object {
      "debut": 1475954335,
      "id": "3",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
  },
  "byPlanID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
  "byToolID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
}
`;

exports[`Tasks Reducer FAIL_TASK should mark the task as failed 1`] = `
Immutable.Record {
  "byTaskID": Immutable.Map {
    "1": Object {
      "debut": 1475954335,
      "id": "1",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "2": Object {
      "busy": false,
      "debut": 1475954335,
      "end": 1475954336,
      "id": "2",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "failed",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "3": Object {
      "debut": 1475954335,
      "id": "3",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
  },
  "byPlanID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
  "byToolID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
}
`;

exports[`Tasks Reducer START_TASK should mark the task as started 1`] = `
Immutable.Record {
  "byTaskID": Immutable.Map {
    "1": Object {
      "debut": 1475954335,
      "id": "1",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "2": Object {
      "debut": 1475954335,
      "end": undefined,
      "id": "2",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "running",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "3": Object {
      "debut": 1475954335,
      "id": "3",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
  },
  "byPlanID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
  "byToolID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
}
`;

exports[`Tasks Reducer STOP_TASK should mark the task as stopped 1`] = `
Immutable.Record {
  "byTaskID": Immutable.Map {
    "1": Object {
      "debut": 1475954335,
      "id": "1",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "/ok",
        "type": "shell",
      },
    },
    "2": Object {
      "busy": false,
      "debut": 1475954335,
      "end": 1475954336,
      "id": "2",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "stopped",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "/ok",
        "type": "shell",
      },
    },
    "3": Object {
      "debut": 1475954335,
      "id": "3",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "/ok",
        "type": "shell",
      },
    },
  },
  "byPlanID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
  "byToolID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
}
`;

exports[`Tasks Reducer STOP_TASK_OF_PLAN should mark the task as stopped 1`] = `
Immutable.Record {
  "byTaskID": Immutable.Map {
    "1": Object {
      "debut": 1475954335,
      "id": "1",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "2": Object {
      "busy": false,
      "debut": 1475954335,
      "end": 1475954336,
      "id": "2",
      "plan": Object {
        "config": 2,
        "id": "2",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "2",
          "name": "tool",
        },
      },
      "state": "stopped",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "3": Object {
      "debut": 1475954335,
      "id": "3",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
  },
  "byPlanID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "3",
    ],
    "2": Immutable.List [
      "2",
    ],
  },
  "byToolID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "3",
    ],
    "2": Immutable.List [
      "2",
    ],
  },
}
`;

exports[`Tasks Reducer SUCCEED_TASK should mark the task as succeed 1`] = `
Immutable.Record {
  "byTaskID": Immutable.Map {
    "1": Object {
      "debut": 1475954335,
      "id": "1",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "2": Object {
      "busy": false,
      "debut": 1475954335,
      "end": 1475954336,
      "id": "2",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "succeed",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "3": Object {
      "debut": 1475954335,
      "id": "3",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
  },
  "byPlanID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
  "byToolID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
}
`;

exports[`Tasks Reducer WAITING_TASK should mark the task as waiting 1`] = `
Immutable.Record {
  "byTaskID": Immutable.Map {
    "1": Object {
      "debut": 1475954335,
      "id": "1",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "2": Object {
      "busy": false,
      "debut": 1475954335,
      "id": "2",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "running",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
    "3": Object {
      "debut": 1475954335,
      "id": "3",
      "plan": Object {
        "config": 2,
        "id": "1",
        "packageInfo": Object {
          "name": "root",
          "path": "/",
          "type": "file",
        },
        "stager": Object {
          "host": null,
          "type": "integrated",
        },
        "tool": Object {
          "iconUri": "",
          "id": "1",
          "name": "tool",
        },
      },
      "state": "created",
      "step": 1,
      "strategy": Object {
        "command": "",
        "cwd": "",
        "type": "shell",
      },
    },
  },
  "byPlanID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
  "byToolID": Immutable.Map {
    "1": Immutable.List [
      "1",
      "2",
      "3",
    ],
  },
}
`;
